crates/ingress-client/src/session.rs

```diff
@@ -0,0 +1,501 @@
+                            batch.committed();
+                        }
+                        Ok(response) => {
+                            // Handle any other error as a connection loss
+                            // and retry all inflight batches.
+                            debug!("Ingest response '{:?}'", response);
+                            break SessionState::Disconnected;
+                        }
+                        Err(_err) => {
+                            // we can assume that for any error
+                            // we need to retry all the inflight bathes.
+                            // special case for load shedding we could
+                            // throttle the stream a little bit then
+                            // speed up over a period of time.
+
+                            break SessionState::Disconnected;
+                        }
+                    }
+                }
+                _ = ctx.cancelled() => {
+                    // relies on auto drain and drop of the inflight
+                    // batches to notify callers that records has been
+                    // cancelled.
+                    return SessionState::Shutdown;
+                }
+            }
+        };
+
+        // don't lose the buffered batch in case of ChunksTimeout
+        if let Either::Left(chunked) = chunked.as_pin_mut() {
+            let remainder = chunked.into_remainder();
+            if !remainder.is_empty() {
+                let (trackers, records): (Vec<_>, Vec<_>) = remainder.into_iter().unzip();
+                let batch: Arc<[IngestRecord]> = Arc::from(records);
+                self.inflight.push_back(IngressBatch {
+                    records: batch,
+                    trackers,
+                    reply_rx: None,
+                });
+            }
+        }
+
+        state
+    }
+}
+
+struct SessionManagerInner<T> {
+    networking: Networking<T>,
+    partition_routing: PartitionRouting,
+    opts: SessionOptions,
+    ctx: CancellationToken,
+    published: ArcSwap<HashMap<PartitionId, SessionHandle>>,
+    locks: DashMap<PartitionId, OnceLock<SessionHandle>>,
+}
+
+impl<T> SessionManagerInner<T>
+where
+    T: TransportConnect,
+{
+    /// Gets or start a new session to partition with given partition id.
+    /// It guarantees that only one session is started per partition id.
+    pub fn get(&self, id: PartitionId) -> SessionHandle {
+        let inner = self.published.load();
+        match inner.get(&id) {
+            Some(handle) => handle.clone(),
+            None => {
+                let once = self.locks.entry(id).or_default();
+
+                let handle = once.get_or_init(|| {
+                    let session = PartitionSession::new(
+                        self.networking.clone(),
+                        self.partition_routing.clone(),
+                        id,
+                        self.opts.clone(),
+                    );
```

Comment:
Why is it important to put a `OnceLock` instead of directly putting a `SessionHandle` value?